mypy/checkexpr.py

```diff
@@ -327,7 +327,20 @@ def infer_arg_types_in_context2(
         for i, actuals in enumerate(formal_to_actual):
             for ai in actuals:
                 if arg_kinds[ai] not in (nodes.ARG_STAR, nodes.ARG_STAR2):
-                    res[ai] = self.accept(args[ai], callee.arg_types[i])
+                    callee_arg_type = callee.arg_types[i]
+                    arg_type = self.accept(args[ai], callee_arg_type)
+                    if (isinstance(arg_type, CallableType) and
```

Comment:
The example I used involved a `[...]` expression in the argument so the list-of-generic-function could be specialized based on context (but isn't specialized correctly), but you are correct and it's a different case due to invariance.